libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,831 @@
+        // don't attempt to configure if we've unread data
+        return;
+    }
+
+    // note that we return as soon as we've transmitted anything in
+    // case we're in one-wire mode
+    for (uint8_t i=0; i<_esc_count; i++) {
+        auto &esc = _escs[i];;
+        switch (esc.state) {
+        case ESCState::WANT_SEND_OK_TO_GET_RUNNING_SW_TYPE:
+            // probe for bootloader or running firmware
+            if (transmit_config_request(PackedMessage<OK>{esc.id, OK{}})) {
+                esc.set_state(ESCState::WAITING_OK_FOR_RUNNING_SW_TYPE);
+            }
+            return;
+        case ESCState::WAITING_OK_FOR_RUNNING_SW_TYPE:
+            return;
+        case ESCState::WANT_SEND_START_FW:
+            if (transmit_config_request(PackedMessage<START_FW>{esc.id, START_FW{}})) {
+                esc.set_state(ESCState::WAITING_OK_FOR_START_FW);
+            }
+            return;
+        case ESCState::WAITING_OK_FOR_START_FW:
+            return;
+#if HAL_AP_FETTEC_ONEWIRE_GET_STATIC_INFO
+        case ESCState::WANT_SEND_REQ_TYPE:
+            if (transmit_config_request(PackedMessage<REQ_TYPE>{esc.id, REQ_TYPE{}})) {
+                esc.set_state(ESCState::WAITING_ESC_TYPE);
+            }
+            return;
+        case ESCState::WAITING_ESC_TYPE:
+            return;
+        case ESCState::WANT_SEND_REQ_SW_VER:
+            if (transmit_config_request(PackedMessage<REQ_SW_VER>{esc.id, REQ_SW_VER{}})) {
+                esc.set_state(ESCState::WAITING_SW_VER);
+            }
+            return;
+        case ESCState::WAITING_SW_VER:
+            return;
+        case ESCState::WANT_SEND_REQ_SN:
+            if (transmit_config_request(PackedMessage<REQ_SN>{esc.id, REQ_SN{}})) {
+                esc.set_state(ESCState::WAITING_SN);
+            }
+            return;
+        case ESCState::WAITING_SN:
+            return;
+#endif
+#if HAL_WITH_ESC_TELEM
+        case ESCState::WANT_SEND_SET_TLM_TYPE:
+            if (transmit_config_request(PackedMessage<SET_TLM_TYPE>{esc.id, SET_TLM_TYPE{1}})) {
+                esc.set_state(ESCState::WAITING_SET_TLM_TYPE_OK);
+            }
+            return;
+        case ESCState::WAITING_SET_TLM_TYPE_OK:
+            return;
+#endif
+        case ESCState::WANT_SEND_SET_FAST_COM_LENGTH: {
+            // FIXME: tidy this up a bit
+            const uint16_t bit_count = 12 + (_esc_count * 11);
+            const uint8_t byte_count = 1 + (bit_count + 7)/8;
+            if (transmit_config_request(PackedMessage<SET_FAST_COM_LENGTH>{esc.id,
+                            SET_FAST_COM_LENGTH{
+                            byte_count,
+                                _escs[0].id,
+                                _esc_count
+                                }})) {
+                esc.set_state(ESCState::WAITING_SET_FAST_COM_LENGTH_OK);
+            }
+        }
+        case ESCState::WAITING_SET_FAST_COM_LENGTH_OK:
+            return;
+        case ESCState::RUNNING: {

```

Comment:
This is the most common case, maybe move it to the top of the switch?
Or is it a stupid micro-optimization? This is after all the 400Hz loop that some day might become the 4KHz loop or something like that

Reply:
I'd be interested to see some analysis of the disassembled binary from, say, a CubeOrange build for this.

Looks like C++20 gives us `[[likely]]` attributes we could apply to ensure the compiler had enough information to make a good decision.